### ISSUE TITLE
Debugging performance issues

### ISSUE THREAD

**Issue by joshas** [fman-community/fman#179]
Sometimes when navigating directories on macOS, I get the "spinning beach ball" cursor as fman stops for a few seconds to do something. It happens after entering a directory - I can see its contents, but I cannot do anything else. After that leaving and entering same directory a few times works normally, without lagging.
Sometimes GoTo dialog lags in a similar way, but that might be a different issue.
Is there any way to debug, what is causing this lag?
Only few plugins are installed and I'm almost certain that they are not related to this.

**Comment by mherrmann** [comment]
The reason it is faster the second time you enter the directory is that fman caches files.

I unfortunately can't think of a specific thing to debug. Do the directories have a lot of files in them? Are they (or do they contain) files / folders that don't lie on your local HDD, but (say) on a network share?

**Comment by joshas** [comment]
In some projects I have node_modules directories, which are notorious for having lots for directories and files inside. How deep fman cache is scanning? Could it be done in background thread?

**Comment by mherrmann** [comment]
It is being done in a background thread. But it's slow before you even look at the `node_modules` directory, right? Or is it slow to render / scroll as well?

**Comment by joshas** [comment]
Here's a common scenario: I press enter on directory to enter it, fman freezes (breach ball spinning icon appears and I'm still inside old directory) and after few moments new directory contents are displayed. Going back up and into same directory after that doesn't cause any delays. This happened on a directory with 6 SVG files and no children directories in it. Might it be that file thumbs are being loaded in background? But fman doesn't use those.
Could you remind me where to find GoTo paths cache file on macOS? I couldn't find it in Library/Application Support directory.

**Comment by kek91** [comment]
Haven't experienced this myself, but regarding your last question; Run `List plugins` from GoTo and open the Settings plugin, this should take you to your Settings-directory which contains `Visited Paths (macOS).json`

**Comment by mherrmann** [comment]
Thx for the explanation Kim (@kek91). Your method is actually quicker than the one I normally use ;) `Open data directory` then navigate down to the plugin ;)

Has it always been that way or has it gotten worse with time @joshas?

**Comment by joshas** [comment]
@mherrmann: Probably month or so, but I cannot pin the issue to some specific release of fman.
Another thing that I noticed, that fman sometime freezes when I return to it after some time e.g. been working in another program and switch to fman - spinning beach ball appears for a second or two.
Visited Paths (macOS).json file weighs only 48 KB with 882 paths, probably not too much to start causing problems?
I'll try to start with fresh install again, maybe some things will improve.
Or is it just hard drive, that's a bit slow?

**Comment by mherrmann** [comment]
I don't think it would be caused by a recent fman release because I don't think anything has changed with regards to reading/displaying directories. It would be interesting to know if things improve with a fresh install (also when you delete / temporarily remove `Visited Paths.json`). Thanks.

**Comment by joshas** [comment]
Tried working today with fresh install - removed all files from Library/Application Support/fman . Still getting same performance issues. I'll try do run some I/O monitor in background to see what's happening behind scenes. Is there any good tools for that on macOS? Default Activity Monitor looks a bit lacking in features.

**Comment by mherrmann** [comment]
I'm sorry, I don't know of any good tools on macOS :/

**Comment by joshas** [comment]
Done some debugging with `sudo fs_usage -f filesys fman` utility, and had no huge freezes all day. Just my luck. Two things that I noticed about fman I/O:
1) fman is doing a lot of fsgetpath and getattrlist calls when cursor is moved to next item and when fman main window is focused or switched to another window.
2) fman is querying info about files not available in current view, e.g. I have ~/work/project/ directory open, but fman is querying ~/work/other-dir/ directory too.
Of course this might just be a total misinterpretation of actual fman operations, I'm just trying to work out what might have caused slowdowns for me.
Finally, thing I forgot to notice, is that during all week one of network drives was very slow. If my deduction about fman reading related directories is correct - that might have caused all slowdowns, as network drive is attached on /Volumes/remote/ and might been "indexed" during directory navigation operations.

 I propose to convert this issue to "improvement" for adding some way to debug fman performance, e.g. start fman with specific flag, to start tracking time taken by basic operations, e.g. startup, enter directory and similar to log file. In case of performance issues, they could easily be tracked down.

**Comment by mherrmann** [comment]
Thank you for your investigation Vaidotas. The slow network drive is a good idea. I have converted this issue to an "enhancement", like you suggested.

Maybe a little context: fman uses something like a "list view" from Qt to display and load files. (To be exact: `QFileSystemModel` and `QTreeView`.) It's these components that do all the fetching and caching of files from the file system, in a background thread. As fman begins to support other file systems (eg. zip files #37), this will have to be replaced by fman's own implementation. Then, I have much more control over which files are fetched when, and can figure out much better why something is slow. For now, it's more like a "black box" that I don't have much control over. Implementing support for other file systems (and thus for archives, FTP etc.) is the next big step in fman's development.

**Comment by raguay** [comment]
I've seen similar slowdowns when looking at remote file systems mounted with CloudMounter. But, fman browses them faster than Path Finder or Finder. Commander One does better, but it supports the remote file systems natively (Same company that makes CoundMounter).

Since I work with a lot of remote file systems, making this a native part of fman would be a great help!

**Comment by mherrmann** [comment]
@raguay I'll keep it in mind!

**Comment by joshas** [comment]
@mherrmann I think I finally got to the bottom of this issue. It seems problems start when there are many remote paths cached in Visited Paths.json file.
Last week opening GoTo dialog became just unbearably long, so I moved Visited Paths.json file and after restart fman was opening GoTo dialog in an instant. Quick look at my Visited Paths.json file revealed, that many different paths were on remote network drive. I have NAS mounted on /Volumes/remote/ and use different directories during work. While there's only a few directories indexed - GoTo opens instantly. I might try to see how many remote paths it takes to slow GoTo dialog down.
So, in conclusion, would it be possible to work around this issue? Detecting remote directories might be impossible, because the way they are mounted. How about some setting, so fman could ignore scaning remote paths when opening GoTo dialog?

**Comment by mherrmann** [comment]
@joshas thanks for this insight. I believe I know why it is taking a long time in this case. When fman displays the GoTo results for a query, it searches through the entries in `Visited Paths.json` and checks if the respective directories still exist. This is to prevent GoTo from suggesting non-existent folders.

The best solution I can think of would be for fman to continuously go through `Visited Paths.json` in the background and purge directories that no longer exist. GoTo would then not have to do this check, and remain fast. It might still sometimes display non-existent directories. But it should be rare and over time, they would disappear. It should be done in a "low-priority" kind of way so no resources are wasted unnecessarily.

An alternative would be to make it possible to "exclude" folders from GoTo. In your case @joshas, you could add `/Volumes/remote` to an exclude list. The drawbacks would be that 1) the user has to know about the setting and 2) you would not be able to "GoTo" any of the directories below `/Volumes/remote`.

The first solution is would probably better right?

**Comment by joshas** [comment]
First solution is better, because user will have best experience without additional tinkering with settings.

**Comment by joshas** [comment]
Any updates on this? Maybe we should create a separate issue for specific feature, so more users could vote on it?

**Comment by mherrmann** [comment]
Are you talking about opening a directory in fman, or about opening GoTo @joshas?

**Comment by joshas** [comment]
Opening GoTo dialog. I seems that after few months of usage it accumulates large count of different directories and takes longer to show up. How about adding "clear cache" function, to clear `Visited Paths.json` right from fman?

**Comment by mherrmann** [comment]
I wonder if a better solution might not be to limit the number of entries in `Visited Paths.json`. What do you get when you do

    python -c 'import json; print(len(json.load(open("Visited Paths (Mac).json"))))'

on the command line in the directory that contains `Visited Paths (Mac).json`, ie. the Settings plugin?

**Comment by joshas** [comment]
Visited Paths (Mac).json contains 1689 entries. About a third of them on remote network drive. Annoying this is that I cannot delete Visited Paths file from fman, as it gets recreated after closing program.
It would be great if GoTo dialog appeared instantly, I could start typing right away and results would be delayed for a bit if directory check is required.

**Comment by mherrmann** [comment]
My `Visited Paths.json` only contains ~300 entries. You said that GoTo did not appear instantly. I suspect this is because it displays the full list of 1689 entries.

To fix this, I have now made it so that `Visited Paths.json` is truncated to 250 once it reaches 500. This ensures that at most 500 entries are displayed. I hope it will make the dialog >3x faster for you.

**Comment by mherrmann** [comment]
What I mentioned in my previous comment is now implemented in [fman 1.4.5](https://fman.io/changelog#v1.4.5). I'm closing this issue. Please open a new one for other performance-related problems.
